### PR TITLE
In ExprRequirement building, treat OverloadExpr as dependent

### LINKED ISSUE #66612: Clang frontend C++ crash with atomic constraints
To quickly reproduce: https://gcc.godbolt.org/z/Gc1Y17PKf (assertion-trunk)

```cpp
#include <concepts>
#include <iostream>

template <typename T>
concept Iterator = requires(T a) {
  { a } -> std::same_as<T>;
};

template <typename T>
concept Container = requires(T a) {
  { std::end } -> Iterator;
};
```

Compiling the above code crashes clang `clang++ -x c++ --std=c++23 `, crashes locally using clang-17.0 (a10019a), also on trunk with assertion (see godbolt link)

### PR DESCRIPTION
**Pull request by erichkeane**
As reported in #66612, we aren't correctly treating the placeholder expression type correctly, so we ended up trying to get a reference version of it, and this resulted in an assertion, since the placeholder type cannot have a reference added.

Fixes: #66612

### CODE REVIEW

File: clang/lib/Sema/SemaExprCXX.cpp [hunk 1]
```diff
@@ -9063,7 +9063,8 @@ Sema::BuildExprRequirement(
     concepts::ExprRequirement::ReturnTypeRequirement ReturnTypeRequirement) {
   auto Status = concepts::ExprRequirement::SS_Satisfied;
   ConceptSpecializationExpr *SubstitutedConstraintExpr = nullptr;
-  if (E->isInstantiationDependent() || ReturnTypeRequirement.isDependent())
+  if (E->isInstantiationDependent() || E->getType()->isPlaceholderType() ||
+      ReturnTypeRequirement.isDependent())
```
Comment (erichkeane) [inline clang/lib/Sema/SemaExprCXX.cpp:9067]: This makes sense in my head, as what is currently an unresolved overload expr(or other overload expr) could potentially become legal after instantiation I think?  So this should result in this being instantiated later/checked for correctness then.

Comment (cor3ntin) [reply clang/lib/Sema/SemaExprCXX.cpp:9067]: I've been trying to think about whether checking `UnresolvedLookupExpr` would make more sense but
I think what you have works in all cases.

**Comment by cor3ntin** [comment]
Do we need a release note for this (probably!)?

**Comment by erichkeane** [comment]
> Do we need a release note for this (probably!)?

Bah, yes we do...

File: clang/lib/Sema/SemaExprCXX.cpp (hunk 1 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaExprCXX.cpp:9067]: I was concerned about other expression types as well, like `OverloadExpr` or placeholders that got put in elsewise, so I was hoping this was sufficiently generic here.

**Review by cor3ntin — APPROVED** [review]
LGTM (but now you have conflicts in the release notes!)

**Pull request merged**